### DISCUSSION TITLE
Mid/Side Encoding/Decoding

### DISCUSSION THREAD

**Discussion by willdatrill2007** [pschatzmann/arduino-audio-tools#1071] [Q&A] [answered]
Is it possible to do Mid/Side Encoding or Decoding? I'm trying to figure out how to invert the phase on one audio channel.

**Comment by pschatzmann** [comment, marked as answer]
I always had it on my todo list to extend the [CallbackStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_callback_stream.html) so that we can easily implement some update methods on the data in addition to the read and write callback methods.
I just committed the related code with a corresponding [test example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/callback/callback-write/callback-write.ino). The update is not only working on the write but also for readBytes...

**Reply by willdatrill2007** [reply]
Ok, Great. Will check it out.

**Comment by willdatrill2007** [comment]
Will this invert the phase on both channels? If yes, how do you invert phase on only one channel?

**Comment by pschatzmann** [comment]
Yes, it's on both channels. If you only need one channel you can process only every other sample: e.g all samples with an even or with an odd index if you have a stereo signal.

    for (int j = 0; j < sample_count; j++) {
      if (j % 2==0) data16[j] = -data16[j]; // invert left channel
    }

**Reply by willdatrill2007** [reply]
Got it. Thank you so much.

**Reply by willdatrill2007** [reply]
I had to change the code for the declaration of the "invert" callback to this and it started to work. I figured it out as I was operating the i2s at 32bit.

size_t sample_count = bytes / sizeof(int32_t);
  int32_t* data32 = (int32_t*)data;
  
Will be testing more.

**Reply by willdatrill2007** [reply]
auto mid = [](uint8_t* data, size_t bytes) {
  size_t sample_count = bytes / sizeof(int32_t);
  int32_t* out_1 = (int32_t*)data;
  for (int k = 0; k < sample_count; k += 2) { // loop over every two samples
    int32_t left = out_1[k]; // get the left channel sample
    int32_t right = out_1[k + 1]; // get the right channel sample
    int32_t mono = (left + right) / 2; // calculate the average of left and right
    out_1[k] = mono; // overwrite the left channel with the mono value
    out_1[k + 1] = mono; // overwrite the right channel with the mono value
  }
  return bytes;
};

auto side = [](uint8_t* data, size_t bytes) {
  size_t sample_count = bytes / sizeof(int32_t);
  int32_t* out_1 = (int32_t*)data;
  for (int k = 0; k < sample_count; k += 2) { // loop over every two samples
    int32_t left = -out_1[k]; // get the left channel sample
    int32_t right = out_1[k + 1]; // get the right channel sample
    int32_t side = (left + right) / 2; // calculate the average of left and right
    out_1[k] = side; // overwrite the left channel with the side value
    out_1[k + 1] = side; // overwrite the right channel with the side value
    out_1[k] = -out_1[k]; // invert one channel phase
  }
  return bytes;
};


I made some progress with my project but when I try to use two callbacks at the same time, one labelled "mid" and the other "side". I sent each to a "VolumeStream" and assign two copiers in the loop but the audio sounds distorted, The volume controls do work. Any assistance would be greatly appreciated.

**Reply by pschatzmann** [reply]
Why would you do this? You can easily do the processing in one call back. if you want to use 2 CallbackStreams, you would need to put them in a chain!

**Reply by willdatrill2007** [reply]
Got it. I realized that I can merge the two into one this morning and it's working great so far @ 96000hz, 32bit, buffer count at 3 and buffer size at 1024. Will keep you informed as I go along. Thanks again for all the assistance.
